server/server.go

```
@@ -854,9 +798,25 @@ func (s *MCPServer) handleToolCall(
-	s.toolsMu.RLock()
-	tool, ok := s.tools[request.Params.Name]
-	s.toolsMu.RUnlock()
+	// First check session-specific tools
+	var tool ServerTool
+	var ok bool
+	
+	session := ClientSessionFromContext(ctx)
+	if session != nil {
+		if sessionWithTools, ok := session.(SessionWithTools); ok {
+			if sessionTools := sessionWithTools.GetSessionTools(); sessionTools != nil {
+				tool, ok = sessionTools[request.Params.Name]
```

golangci-lint (1.64.8)

ineffectual assignment to ok

(ineffassign)